### PR TITLE
ENH: Add option to set Segment Editor threshold values from a local ROI

### PR DESCRIPTION
**Pull request by Sunderlandkyl**
The "Local histogram" section within the Segment Editor threshold effect allows users to select a brush and click+drag on the image to create an ROI

Available brushes are:
  * Box
  * Circle
  * Draw
  * Line

Values within the ROI are plotted on a histogram, and the upper and lower threshold values are set based on the maximum, minimum or average value from the histogram.

Clicking and dragging on the histogram will create a manually selected range that will be used to select the threshold values. Right-clicking will clear the manually selected range.

image

Requires https://github.com/commontk/CTK/pull/888

### CODE REVIEW

File: Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorThresholdEffect.py [hunk 1]
```diff
@@ -444,6 +633,206 @@ def preview(self):
...
+    elif eventId == vtk.vtkCommand.LeftButtonReleaseEvent:
+      self.histogramPipeline.state = HISTOGRAM_STATE_PLACED
+      abortEvent = True
+    elif eventId == vtk.vtkCommand.MouseMoveEvent:
+      if self.histogramPipeline.state == HISTOGRAM_STATE_MOVING:
+        self.histogramPipeline.addPoint(ras)
+        self.updateHistogram()
+        abortEvent = True
+    else:
+      pass
```
Comment (lassoan) [inline Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorThresholdEffect.py]: not needed

File: Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorThresholdEffect.py [hunk 2]
```diff
@@ -444,6 +633,206 @@ def preview(self):
...
+    brushExtent = [0, -1, 0, -1, 0, -1]
+    for i in range(3):
+      brushExtent[2*i] = vtk.vtkMath.Floor(brushBounds[2*i])
+      brushExtent[2*i+1] = vtk.vtkMath.Ceil(brushBounds[2*i+1])
+    if brushExtent[0] > brushExtent[1] or brushExtent[2] > brushExtent[3] or brushExtent[4] > brushExtent[5]:
+      self.histogramFunction.RemoveAllPoints()
+      return
+
+    sliceLogic = self.histogramPipeline.sliceWidget.sliceLogic()
+    backgroundLogic = sliceLogic.GetBackgroundLayer()
```
Comment (lassoan) [inline Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorThresholdEffect.py]: Usually the master volume is the background volume but the user can change that. Maybe we could choose between foreground/background depending on which one is the master volume and if neither then log a warning and use the more visible one.

File: Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorThresholdEffect.py [hunk 3]
```diff
@@ -444,6 +633,206 @@ def preview(self):
...
+    self.averageFunction.RemoveAllPoints()
+    self.averageFunction.AddPoint(average - epsilon, 0.0)
+    self.averageFunction.AddPoint(average, 1.0)
+    self.averageFunction.AddPoint(average + epsilon, 0.0)
+    self.averageFunction.AdjustRange(scalarRange)
+
+    minimumThreshold = lower
+    maximumThreshold = upper
+
+    if self.histogramLowerThresholdMinimumButton.checked:
```
Comment (lassoan) [inline Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorThresholdEffect.py]: it would be a bit nicer to use parameter value in self.scriptedEffect node instead of GUI button state

File: Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorThresholdEffect.py [hunk 4]
```diff
@@ -482,6 +871,251 @@ def __init__(self):
...
+
+    self.rasPoints = vtk.vtkPoints()
+    self.polyData = vtk.vtkPolyData()
+    self.polyData.SetPoints(self.rasPoints)
+
+    lines = vtk.vtkCellArray()
+    self.polyData.SetLines(lines)
+    idArray = lines.GetData()
+    idArray.Reset()
+    idArray.InsertNextTuple1(0)
```
Comment (lassoan) [inline Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorThresholdEffect.py]: why is this needed?

**Review by lassoan — COMMENTED** [review]
Thank you, this is a great feature! Please implement saving/restoring of user choices (ROI shape, lower/upper value selection) on the GUI to/from parameter node. I've also added a few minor comments.

File: Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorThresholdEffect.py (hunk 2 above)
Comment (Sunderlandkyl) [reply Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorThresholdEffect.py]: Sure. Currently the background is used for the threshold preview. Should that be changed as well?

Comment (lassoan) [reply Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorThresholdEffect.py]: Yes, it would be nice to use the same logic for both.

**Comment by lassoan** [comment]
Thank you, merged!
